npm/private/npm_link_package_store.bzl

```diff
@@ -0,0 +1,93 @@
+
+The term "package" is defined at
+<https://nodejs.org/docs/latest-v16.x/api/packages.html>
+
+See https://pnpm.io/symlinked-node-modules-structure for more information on
+the symlinked node_modules structure.
+Npm may also support a symlinked node_modules structure called
+"Isolated mode" in the future:
+https://github.com/npm/rfcs/blob/main/accepted/0042-isolated-mode.md.
+"""
+
+_ATTRS = {
+    "src": attr.label(
+        doc = """The npm_package_store target to link as a direct dependency.""",
+        providers = [NpmPackageStoreInfo],
+        mandatory = True,
+    ),
+    "package": attr.string(
+        doc = """The package name to link to.
+
+If unset, the package name of the src npm_package_store is used.
+If set, takes precendance over the package name in the src npm_package_store.
+""",
+    ),
+    "allow_unresolved_symlinks": attr.bool(
+        mandatory = True,
+        doc = """Whether unresolved symlinks are enabled in the current build configuration.
+
+        These are enabled with the --experimental_allow_unresolved_symlinks flag.
+
+        Typical usage of this rule is via a macro which automatically sets this
+        attribute based on a `config_setting` rule.
+        """,
+    ),
+}
+
+def _impl(ctx):
+    virtual_store_directory = ctx.attr.src[NpmPackageStoreInfo].virtual_store_directory
+    if not virtual_store_directory:
+        fail("src must be a npm_link_package that provides a virtual_store_directory")
+
+    package = ctx.attr.package if ctx.attr.package else ctx.attr.src[NpmPackageStoreInfo].package
+
+    # symlink the package's path in the virtual store to the root of the node_modules
+    # "node_modules/{package}" so it is available as a direct dependency
+    root_symlink_path = paths.join("node_modules", package)
+
+    files = utils.make_symlink(ctx, root_symlink_path, virtual_store_directory)
+    files_depset = depset(files)
+
+    transitive_files_depset = depset(files, transitive = [ctx.attr.src[NpmPackageStoreInfo].transitive_files])
+
+    providers = [
+        DefaultInfo(
+            files = transitive_files_depset,
+        ),
+        NpmLinkedPackageInfo(
+            label = ctx.label,
+            link_package = ctx.label.package,
+            package = ctx.attr.src[NpmPackageStoreInfo].package,
+            version = ctx.attr.src[NpmPackageStoreInfo].version,
+            store_info = ctx.attr.src[NpmPackageStoreInfo],
+            files = files_depset,
+            transitive_files = transitive_files_depset,
+        ),
+        DeclarationInfo(
+            declarations = depset(files, transitive = [ctx.attr.src[DeclarationInfo].declarations]),
+            transitive_declarations = depset(files, transitive = [ctx.attr.src[DeclarationInfo].transitive_declarations]),
+        ),
+    ]
+    if OutputGroupInfo in ctx.attr.src:
+        providers.append(ctx.attr.src[OutputGroupInfo])
+
+    return providers
```

Comment:
we could have some analysis unit tests for the providers being right-shaped, seem useful?

Reply:
yup; those are useful. lets do as a follow-up